File: libraries/python/src/edugraph/relations.py
"""Maintained compatibility adapters over the immutable bundled context."""

from collections.abc import Iterable
from functools import cache
from importlib.resources import files
from typing import TypeAlias, TypedDict

from .ability import Ability
from .area import Area
from .core import OntologyContext, snapshot_from_json
from .scope import Scope

CompetencyDescriptor: TypeAlias = Area | Scope | Ability


class DescriptorRelations(TypedDict, total=False):
    """Detached legacy relation lists and the flattened release definition."""

    definition: str
    structures: list[CompetencyDescriptor]
    structuredBy: list[CompetencyDescriptor]
    partOf: list[CompetencyDescriptor]
    hasPart: list[CompetencyDescriptor]
    specializes: list[CompetencyDescriptor]
    specializedBy: list[CompetencyDescriptor]
    constrains: list[CompetencyDescriptor]
    constrainedBy: list[CompetencyDescriptor]
    implies: list[CompetencyDescriptor]
    impliedBy: list[CompetencyDescriptor]
    contradicts: list[CompetencyDescriptor]
    contradictedBy: list[CompetencyDescriptor]
    expands: list[CompetencyDescriptor]
    expandedBy: list[CompetencyDescriptor]
    inverts: list[CompetencyDescriptor]
    invertedBy: list[CompetencyDescriptor]
    integrates: list[CompetencyDescriptor]
    integratedBy: list[CompetencyDescriptor]
    translates: list[CompetencyDescriptor]
    translatedBy: list[CompetencyDescriptor]


@cache
def bundled_context() -> OntologyContext:
    """Return the shared read-only release context, initialized on first use."""
    text = files(__package__).joinpath("snapshot.json").read_text(encoding="utf-8")
    return OntologyContext(snapshot_from_json(text))


_MEMBERS: dict[str, CompetencyDescriptor] = {
    str(member): member for cls in (Area, Scope, Ability) for member in cls
}


def _members(iris: Iterable[str]) -> list[CompetencyDescriptor]:
    return [_MEMBERS[iri] for iri in iris if iri in _MEMBERS]


def definition(descriptor: CompetencyDescriptor) -> str:
    """Return the released enum definition, or an empty string for unknown input."""
    member = _MEMBERS.get(str(descriptor))
    return member.definition if member is not None else ""


def relations(descriptor: CompetencyDescriptor) -> DescriptorRelations:
    """Return detached relation lists; mutation never changes the bundled ontology."""
    if str(descriptor) not in _MEMBERS:
        return {}
    result: DescriptorRelations = {}
    text = definition(descriptor)
    if text:
        result["definition"] = text
    if targets := structures(descriptor):
        result["structures"] = targets
    if targets := structured_by(descriptor):
        result["structuredBy"] = targets
    if targets := part_of(descriptor):
        result["partOf"] = targets
    if targets := has_part(descriptor):
        result["hasPart"] = targets
    if targets := specializes(descriptor):
        result["specializes"] = targets
    if targets := specialized_by(descriptor):
        result["specializedBy"] = targets
    if targets := constrains(descriptor):
        result["constrains"] = targets
    if targets := constrained_by(descriptor):
        result["constrainedBy"] = targets
    if targets := implies(descriptor):
        result["implies"] = targets
    if targets := implied_by(descriptor):
        result["impliedBy"] = targets
    if targets := contradicts(descriptor):
        result["contradicts"] = targets
    if targets := contradicted_by(descriptor):
        result["contradictedBy"] = targets
    if targets := expands(descriptor):
        result["expands"] = targets
    if targets := expanded_by(descriptor):
        result["expandedBy"] = targets
    if targets := inverts(descriptor):
        result["inverts"] = targets
    if targets := inverted_by(descriptor):
        result["invertedBy"] = targets
    if targets := integrates(descriptor):
        result["integrates"] = targets
    if targets := integrated_by(descriptor):
        result["integratedBy"] = targets
    if targets := translates(descriptor):
        result["translates"] = targets
    if targets := translated_by(descriptor):
        result["translatedBy"] = targets
    return result


def structures(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released structures targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#structures"))


def structures_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable structures targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#structures")
    )


def structured_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released structuredBy targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#structuredBy")
    )


def structured_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable structuredBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#structuredBy")
    )


def part_of(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released partOf targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#partOf"))


def part_of_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable partOf targets; self only when reached by a cycle."""
    return _members(bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#partOf"))


def has_part(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released hasPart targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#hasPart"))


def has_part_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable hasPart targets; self only when reached by a cycle."""
    return _members(bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#hasPart"))


def specializes(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released specializes targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#specializes")
    )


def specializes_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable specializes targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#specializes")
    )


def specialized_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released specializedBy targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#specializedBy")
    )


def specialized_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable specializedBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#specializedBy")
    )


def constrains(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released constrains targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#constrains"))


def constrains_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable constrains targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#constrains")
    )


def constrained_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released constrainedBy targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#constrainedBy")
    )


def constrained_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable constrainedBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#constrainedBy")
    )


def implies(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released implies targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#implies"))


def implies_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable implies targets; self only when reached by a cycle."""
    return _members(bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#implies"))


def implied_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released impliedBy targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#impliedBy"))


def implied_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable impliedBy targets; self only when reached by a cycle."""
    return _members(bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#impliedBy"))


def contradicts(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released contradicts targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#contradicts")
    )


def contradicts_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable contradicts targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#contradicts")
    )


def contradicted_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released contradictedBy targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#contradictedBy")
    )


def contradicted_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable contradictedBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#contradictedBy")
    )


def expands(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released expands targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#expands"))


def expands_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable expands targets; self only when reached by a cycle."""
    return _members(bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#expands"))


def expanded_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released expandedBy targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#expandedBy"))


def expanded_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable expandedBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#expandedBy")
    )


def inverts(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released inverts targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#inverts"))


def inverts_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable inverts targets; self only when reached by a cycle."""
    return _members(bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#inverts"))


def inverted_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released invertedBy targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#invertedBy"))


def inverted_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable invertedBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#invertedBy")
    )


def integrates(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released integrates targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#integrates"))


def integrates_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable integrates targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#integrates")
    )


def integrated_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released integratedBy targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#integratedBy")
    )


def integrated_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable integratedBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#integratedBy")
    )


def translates(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released translates targets, including schema-derived access."""
    return _members(bundled_context().related(str(descriptor), "http://edugraph.io/edu#translates"))


def translates_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable translates targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#translates")
    )


def translated_by(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted released translatedBy targets, including schema-derived access."""
    return _members(
        bundled_context().related(str(descriptor), "http://edugraph.io/edu#translatedBy")
    )


def translated_by_transitive(descriptor: CompetencyDescriptor) -> list[CompetencyDescriptor]:
    """Return sorted reachable translatedBy targets; self only when reached by a cycle."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#translatedBy")
    )


def transitive_closure(
    descriptor: CompetencyDescriptor, relation: str
) -> list[CompetencyDescriptor]:
    """Legacy traversal by local property name; unknown properties return []."""
    return _members(
        bundled_context().traverse(str(descriptor), "http://edugraph.io/edu#" + relation)
    )


def incompatible(a: CompetencyDescriptor, b: CompetencyDescriptor) -> bool:
    """Check recorded implication/contradiction paths in the bundled release."""
    return bundled_context().incompatible(str(a), str(b))


def deduct_compatible(
    base_constraints: Iterable[CompetencyDescriptor],
) -> list[CompetencyDescriptor]:
    """Return existing conjunctive deduction results as a detached sorted list."""
    return _members(bundled_context().deduct_compatible(str(x) for x in base_constraints))


def deduct_admitting(boundaries: Iterable[CompetencyDescriptor]) -> list[CompetencyDescriptor]:
    """Return existing disjunctive boundary results as a detached sorted list."""
    return _members(bundled_context().deduct_admitting(str(x) for x in boundaries))


def is_label_eligible(descriptor: CompetencyDescriptor) -> bool:
    """Apply ONT-E7; unknown runtime input raises UnknownDescriptorError."""
    return bundled_context().is_label_eligible(str(descriptor))


# Detached compatibility snapshots. They are not an editing API.
definitions: dict[CompetencyDescriptor, str] = {
    x: definition(x) for x in _MEMBERS.values() if definition(x)
}
ENTITY_RELATIONS: dict[CompetencyDescriptor, DescriptorRelations] = {
    x: relations(x) for x in _MEMBERS.values()
}

File: libraries/typescript/Relations.ts
import { Area } from "./Area";
import { Scope } from "./Scope";
import { Ability } from "./Ability";
import { createOntologyContext, InvolvementStatementOptions } from "./core";
import { bundledStatements } from "./BundledSnapshot";

/** Released descriptor IRIs. Supplied snapshots can contain additional IRIs. */
export type CompetencyDescriptor = Area | Scope | Ability;
/** One release-bound context used by all bundled-data convenience helpers. */
export const bundledContext = createOntologyContext(bundledStatements);
const allDescriptors: readonly CompetencyDescriptor[] = [...Object.values(Area), ...Object.values(Scope), ...Object.values(Ability)];
const known = new Set<string>(allDescriptors);
function descriptors(iris: readonly string[]): CompetencyDescriptor[] {
  return iris.filter((iri): iri is CompetencyDescriptor => known.has(iri));
}
/** Direct relation view, including declared inverse and superproperty access. */
export interface DescriptorRelations {
  definition?: string;
  structures?: CompetencyDescriptor[];
  structuredBy?: CompetencyDescriptor[];
  partOf?: CompetencyDescriptor[];
  hasPart?: CompetencyDescriptor[];
  specializes?: CompetencyDescriptor[];
  specializedBy?: CompetencyDescriptor[];
  expands?: CompetencyDescriptor[];
  expandedBy?: CompetencyDescriptor[];
  integrates?: CompetencyDescriptor[];
  integratedBy?: CompetencyDescriptor[];
  inverts?: CompetencyDescriptor[];
  invertedBy?: CompetencyDescriptor[];
  translates?: CompetencyDescriptor[];
  translatedBy?: CompetencyDescriptor[];
  constrains?: CompetencyDescriptor[];
  constrainedBy?: CompetencyDescriptor[];
  implies?: CompetencyDescriptor[];
  impliedBy?: CompetencyDescriptor[];
  contradicts?: CompetencyDescriptor[];
  contradictedBy?: CompetencyDescriptor[];
}
/** Lexically first released definition, or empty text when unavailable. */
export function definition(descriptor: CompetencyDescriptor): string {
  return bundledContext.lookupDescriptor(descriptor)?.definitions[0] ?? "";
}
/** Combine a released descriptor's label, definition, and optional comment for display. */
export function involvementStatement(descriptor: CompetencyDescriptor, options: InvolvementStatementOptions = {}): string {
  return bundledContext.involvementStatement(descriptor, options);
}
/** Fresh relation view. Mutating it cannot modify the bundled ontology context. */
export function relations(descriptor: CompetencyDescriptor): DescriptorRelations {
  const result: DescriptorRelations = {};
  const text = definition(descriptor);
  if (text) result.definition = text;
  const structuresValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#structures"));
  if (structuresValues.length) result.structures = structuresValues;
  const structuredByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#structuredBy"));
  if (structuredByValues.length) result.structuredBy = structuredByValues;
  const partOfValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#partOf"));
  if (partOfValues.length) result.partOf = partOfValues;
  const hasPartValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#hasPart"));
  if (hasPartValues.length) result.hasPart = hasPartValues;
  const specializesValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#specializes"));
  if (specializesValues.length) result.specializes = specializesValues;
  const specializedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#specializedBy"));
  if (specializedByValues.length) result.specializedBy = specializedByValues;
  const expandsValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#expands"));
  if (expandsValues.length) result.expands = expandsValues;
  const expandedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#expandedBy"));
  if (expandedByValues.length) result.expandedBy = expandedByValues;
  const integratesValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#integrates"));
  if (integratesValues.length) result.integrates = integratesValues;
  const integratedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#integratedBy"));
  if (integratedByValues.length) result.integratedBy = integratedByValues;
  const invertsValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#inverts"));
  if (invertsValues.length) result.inverts = invertsValues;
  const invertedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#invertedBy"));
  if (invertedByValues.length) result.invertedBy = invertedByValues;
  const translatesValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#translates"));
  if (translatesValues.length) result.translates = translatesValues;
  const translatedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#translatedBy"));
  if (translatedByValues.length) result.translatedBy = translatedByValues;
  const constrainsValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#constrains"));
  if (constrainsValues.length) result.constrains = constrainsValues;
  const constrainedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#constrainedBy"));
  if (constrainedByValues.length) result.constrainedBy = constrainedByValues;
  const impliesValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#implies"));
  if (impliesValues.length) result.implies = impliesValues;
  const impliedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#impliedBy"));
  if (impliedByValues.length) result.impliedBy = impliedByValues;
  const contradictsValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#contradicts"));
  if (contradictsValues.length) result.contradicts = contradictsValues;
  const contradictedByValues = descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#contradictedBy"));
  if (contradictedByValues.length) result.contradictedBy = contradictedByValues;
  return result;
}
// Every enum value is included, so these compatibility records cover the descriptor union.
/** Released relation views; helper queries always use the isolated core context. */
export const ENTITY_RELATIONS = Object.fromEntries(allDescriptors.map(iri => [iri, relations(iri)])) as Record<CompetencyDescriptor, DescriptorRelations>;
/** Released definition text by descriptor IRI. */
export const definitions = Object.fromEntries(allDescriptors.map(iri => [iri, definition(iri)])) as Record<CompetencyDescriptor, string>;
/** Relation names accepted by the released-data traversal adapter. */
export type RelationKeys = Exclude<keyof DescriptorRelations, "definition">;
/** Reachable unique IRIs, sorted by IRI; self appears only when a cycle reaches it. */
export function transitiveClosure(descriptor: CompetencyDescriptor, relation: RelationKeys): CompetencyDescriptor[] {
  return descriptors(bundledContext.traverse(descriptor, "http://edugraph.io/edu#" + relation));
}
/** Structural eligibility only; unknown runtime IRIs throw instead of receiving a positive answer. */
export function isLabelEligible(descriptor: CompetencyDescriptor): boolean { return bundledContext.isLabelEligible(descriptor); }
/** Recorded implication/contradiction incompatibility; shared range endpoints do not override exclusions. */
export function incompatible(a: CompetencyDescriptor, b: CompetencyDescriptor): boolean { return bundledContext.incompatible(a, b); }
/** Existing conjunctive compatible-label deduction via the shared core. */
export function deductCompatible(constraints: CompetencyDescriptor[]): CompetencyDescriptor[] { return descriptors(bundledContext.deductCompatible(constraints)); }
/** Existing disjunctive boundary-admission deduction via the shared core. */
export function deductAdmitting(boundaries: CompetencyDescriptor[]): CompetencyDescriptor[] { return descriptors(bundledContext.deductAdmitting(boundaries)); }
/** Direct structures targets in the bundled snapshot. */
export function structures(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#structures")); }
/** Reachable structures targets; traversal introduces no additional inference. */
export function structuresTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "structures"); }
/** Direct structuredBy targets in the bundled snapshot. */
export function structuredBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#structuredBy")); }
/** Reachable structuredBy targets; traversal introduces no additional inference. */
export function structuredByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "structuredBy"); }
/** Direct partOf targets in the bundled snapshot. */
export function partOf(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#partOf")); }
/** Reachable partOf targets; traversal introduces no additional inference. */
export function partOfTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "partOf"); }
/** Direct hasPart targets in the bundled snapshot. */
export function hasPart(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#hasPart")); }
/** Reachable hasPart targets; traversal introduces no additional inference. */
export function hasPartTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "hasPart"); }
/** Direct specializes targets in the bundled snapshot. */
export function specializes(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#specializes")); }
/** Reachable specializes targets; traversal introduces no additional inference. */
export function specializesTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "specializes"); }
/** Direct specializedBy targets in the bundled snapshot. */
export function specializedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#specializedBy")); }
/** Reachable specializedBy targets; traversal introduces no additional inference. */
export function specializedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "specializedBy"); }
/** Direct expands targets in the bundled snapshot. */
export function expands(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#expands")); }
/** Reachable expands targets; traversal introduces no additional inference. */
export function expandsTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "expands"); }
/** Direct expandedBy targets in the bundled snapshot. */
export function expandedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#expandedBy")); }
/** Reachable expandedBy targets; traversal introduces no additional inference. */
export function expandedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "expandedBy"); }
/** Direct integrates targets in the bundled snapshot. */
export function integrates(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#integrates")); }
/** Reachable integrates targets; traversal introduces no additional inference. */
export function integratesTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "integrates"); }
/** Direct integratedBy targets in the bundled snapshot. */
export function integratedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#integratedBy")); }
/** Reachable integratedBy targets; traversal introduces no additional inference. */
export function integratedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "integratedBy"); }
/** Direct inverts targets in the bundled snapshot. */
export function inverts(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#inverts")); }
/** Reachable inverts targets; traversal introduces no additional inference. */
export function invertsTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "inverts"); }
/** Direct invertedBy targets in the bundled snapshot. */
export function invertedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#invertedBy")); }
/** Reachable invertedBy targets; traversal introduces no additional inference. */
export function invertedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "invertedBy"); }
/** Direct translates targets in the bundled snapshot. */
export function translates(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#translates")); }
/** Reachable translates targets; traversal introduces no additional inference. */
export function translatesTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "translates"); }
/** Direct translatedBy targets in the bundled snapshot. */
export function translatedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#translatedBy")); }
/** Reachable translatedBy targets; traversal introduces no additional inference. */
export function translatedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "translatedBy"); }
/** Direct constrains targets in the bundled snapshot. */
export function constrains(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#constrains")); }
/** Reachable constrains targets; traversal introduces no additional inference. */
export function constrainsTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "constrains"); }
/** Direct constrainedBy targets in the bundled snapshot. */
export function constrainedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#constrainedBy")); }
/** Reachable constrainedBy targets; traversal introduces no additional inference. */
export function constrainedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "constrainedBy"); }
/** Direct implies targets in the bundled snapshot. */
export function implies(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#implies")); }
/** Reachable implies targets; traversal introduces no additional inference. */
export function impliesTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "implies"); }
/** Direct impliedBy targets in the bundled snapshot. */
export function impliedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#impliedBy")); }
/** Reachable impliedBy targets; traversal introduces no additional inference. */
export function impliedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "impliedBy"); }
/** Direct contradicts targets in the bundled snapshot. */
export function contradicts(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#contradicts")); }
/** Reachable contradicts targets; traversal introduces no additional inference. */
export function contradictsTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "contradicts"); }
/** Direct contradictedBy targets in the bundled snapshot. */
export function contradictedBy(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return descriptors(bundledContext.related(descriptor, "http://edugraph.io/edu#contradictedBy")); }
/** Reachable contradictedBy targets; traversal introduces no additional inference. */
export function contradictedByTransitive(descriptor: CompetencyDescriptor): CompetencyDescriptor[] { return transitiveClosure(descriptor, "contradictedBy"); }
